backend/Watchdog.AspNetCore/WatchdogAspNetMiddleware.cs

```diff
@@ -0,0 +1,66 @@
+﻿using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+using System;
+using System.Threading.Tasks;
+
+namespace Watchdog.AspNetCore
+{
+    public class WatchdogAspNetMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly WatchdogMiddlewareSettings _middlewareSettings;
+        private readonly WatchdogSettings _settings;
+
+        public WatchdogAspNetMiddleware(RequestDelegate next, IOptions<WatchdogSettings> settings, WatchdogMiddlewareSettings middlewareSettings)
+        {
+            _next = next;
+            _middlewareSettings = middlewareSettings;
+            _settings = _middlewareSettings.ClientProvider.GetWatchdogSettings(settings.Value ?? new WatchdogSettings());
+        }
+
+        public async Task InvokeAsync(HttpContext httpContext)
+        {
+            try
+            {
+                await _next(httpContext);
+            }
+            catch(Exception e)
+            {
+                var client = _middlewareSettings.ClientProvider.GetClient(_settings, httpContext);
+                await client.SendInBackground(e);
+                throw;
+            }
+        }
+    }
+
+    public static class ApplicationBuilderExtensions
+    {
```

Comment:
Yeah, that's an important rule, but in this case, I decided to put the two classes in one file because they are tightly coupled and cannot be used separately anywhere in the library.

Reply:
Your words make sense, but as for me 2 classes in one file impairs the readability of the code. For example, I did not immediately notice that there are 2 classes and thought why there are these methods.